File: src/negotiate/handler.js
/**
 * Shared Accept-negotiation + agent-friendly 404.
 * `has` / `read` are injected so the same logic runs in Node (Vite) and Deno (Netlify).
 */

import {
  HTML_TYPE,
  MARKDOWN_TYPE,
  PRODUCES,
  preferredType,
  varyAcceptValue,
  notAcceptableBody,
} from "./accept.js";
import {
  htmlCandidates,
  markdownCandidates,
  isPassthroughPath,
  normalizePath,
} from "./resolve.js";

const MD_CONTENT_TYPE = "text/markdown; charset=utf-8";
const HTML_CONTENT_TYPE = "text/html; charset=utf-8";
const PLAIN_CONTENT_TYPE = "text/plain; charset=utf-8";

/**
 * @typedef {{
 *   has: (path: string) => boolean | Promise<boolean>,
 *   read: (path: string) => string | Promise<string>,
 * }} Store
 */

/**
 * @typedef {{
 *   status: number,
 *   headers: Record<string, string>,
 *   body: string,
 * }} Negotiated
 */

function headersFor(contentType, extra = {}) {
  return {
    "Content-Type": contentType,
    Vary: varyAcceptValue(null),
    ...extra,
  };
}

async function firstExisting(store, candidates) {
  for (const path of candidates) {
    if (await store.has(path)) return path;
  }
  return null;
}

/**
 * @param {{ pathname: string, accept: string | null | undefined }} req
 * @param {Store} store
 * @returns {Promise<Negotiated | null>} null = let the static server handle it
 */
export async function negotiateRequest({ pathname, accept }, store) {
  const clean = normalizePath(pathname);
  if (isPassthroughPath(clean)) return null;

  const chosen = preferredType(accept, PRODUCES);
  if (chosen === null) {
    return {
      status: 406,
      headers: {
        ...headersFor(PLAIN_CONTENT_TYPE),
        "Cache-Control": "no-store",
      },
      body: notAcceptableBody(accept || ""),
    };
  }

  const htmlPath = await firstExisting(store, htmlCandidates(clean));
  const mdPath = await firstExisting(store, markdownCandidates(clean));
  const found = Boolean(htmlPath || mdPath);

  if (!found) {
    return notFoundResponse(chosen, store);
  }

  if (chosen === MARKDOWN_TYPE) {
    if (mdPath) {
      const body = await store.read(mdPath);
      return {
        status: 200,
        headers: headersFor(MD_CONTENT_TYPE, {
          Link: `<${mdPath}>; rel="alternate"; type="text/markdown", </llms.txt>; rel="describedby"`,
        }),
        body,
      };
    }
    if (!preferredType(accept, [HTML_TYPE])) {
      return {
        status: 406,
        headers: {
          ...headersFor(PLAIN_CONTENT_TYPE),
          "Cache-Control": "no-store",
        },
        body: notAcceptableBody(accept || ""),
      };
    }
  }

  if (htmlPath) {
    const body = await store.read(htmlPath);
    const extra = {};
    if (mdPath) {
      extra.Link = `<${mdPath}>; rel="alternate"; type="text/markdown", </llms.txt>; rel="describedby"`;
    }
    return {
      status: 200,
      headers: headersFor(HTML_CONTENT_TYPE, extra),
      body,
    };
  }

  if (mdPath) {
    const body = await store.read(mdPath);
    return {
      status: 200,
      headers: headersFor(MD_CONTENT_TYPE),
      body,
    };
  }

  return null;
}

/**
 * @param {string} chosen
 * @param {Store} store
 */
export async function notFoundResponse(chosen, store) {
  if (chosen === MARKDOWN_TYPE && (await store.has("/404.md"))) {
    return {
      status: 404,
      headers: headersFor(MD_CONTENT_TYPE),
      body: await store.read("/404.md"),
    };
  }
  if (await store.has("/404.html")) {
    return {
      status: 404,
      headers: headersFor(HTML_CONTENT_TYPE),
      body: await store.read("/404.html"),
    };
  }
  const fallback = [
    "# Not found",
    "",
    "This path does not exist on scull7.com.",
    "",
    "## Where to go next",
    "",
    "- [Home](https://scull7.com/)",
    "- [About](https://scull7.com/about)",
    "- [Contact](https://scull7.com/contact)",
    "- [Privacy](https://scull7.com/privacy)",
    "- [Sitemap](https://scull7.com/sitemap.xml)",
    "- [llms.txt](https://scull7.com/llms.txt)",
    "",
  ].join("\n");
  return {
    status: 404,
    headers: headersFor(
      chosen === MARKDOWN_TYPE ? MD_CONTENT_TYPE : PLAIN_CONTENT_TYPE,
    ),
    body: fallback,
  };
}

File: src/melange/handler.ml
(* Shared Accept-negotiation + agent-friendly 404.
   `has` / `read` are injected so Vite preview and tests share this logic. *)

type store = { has : string -> bool; read : string -> string }

type negotiated = {
  status : int;
  headers : (string * string) list;
  body : string;
}

let md_content_type = "text/markdown; charset=utf-8"
let html_content_type = "text/html; charset=utf-8"
let plain_content_type = "text/plain; charset=utf-8"

let headers_for content_type extra =
  ("Content-Type", content_type)
  :: ("Vary", Accept.vary_accept_value None)
  :: extra

let first_existing store candidates =
  let rec loop = function
    | [] -> None
    | path :: rest -> if store.has path then Some path else loop rest
  in
  loop candidates

let not_found_fallback =
  String.concat "\n"
    [
      "# Not found";
      "";
      "This path does not exist on scull7.com.";
      "";
      "## Where to go next";
      "";
      "- [Home](https://scull7.com/)";
      "- [About](https://scull7.com/about)";
      "- [Contact](https://scull7.com/contact)";
      "- [Privacy](https://scull7.com/privacy)";
      "- [Sitemap](https://scull7.com/sitemap.xml)";
      "- [llms.txt](https://scull7.com/llms.txt)";
      "";
    ]

let not_found_response chosen store =
  if chosen = Accept.markdown_type && store.has "/404.md" then
    {
      status = 404;
      headers = headers_for md_content_type [];
      body = store.read "/404.md";
    }
  else if store.has "/404.html" then
    {
      status = 404;
      headers = headers_for html_content_type [];
      body = store.read "/404.html";
    }
  else
    {
      status = 404;
      headers =
        headers_for
          (if chosen = Accept.markdown_type then md_content_type
           else plain_content_type)
          [];
      body = not_found_fallback;
    }

let requested accept = match accept with Some s -> s | None -> ""

let html_result store html_path md_path =
  match html_path with
  | Some path ->
      let extra =
        match md_path with
        | Some md ->
            [
              ( "Link",
                "<" ^ md
                ^ ">; rel=\"alternate\"; type=\"text/markdown\", </llms.txt>; rel=\"describedby\""
              );
            ]
        | None -> []
      in
      Some
        {
          status = 200;
          headers = headers_for html_content_type extra;
          body = store.read path;
        }
  | None -> (
      match md_path with
      | Some path ->
          Some
            {
              status = 200;
              headers = headers_for md_content_type [];
              body = store.read path;
            }
      | None -> None)

let negotiate_request ~pathname ~accept store =
  let clean = Resolve.normalize_path pathname in
  if Resolve.is_passthrough_path clean then None
  else
    match Accept.preferred_type accept Accept.produces with
    | None ->
        Some
          {
            status = 406;
            headers =
              headers_for plain_content_type [ ("Cache-Control", "no-store") ];
            body = Accept.not_acceptable_body (requested accept);
          }
    | Some chosen ->
        let html_path = first_existing store (Resolve.html_candidates clean) in
        let md_path = first_existing store (Resolve.markdown_candidates clean) in
        let found =
          match (html_path, md_path) with None, None -> false | _ -> true
        in
        if not found then Some (not_found_response chosen store)
        else if chosen = Accept.markdown_type then
          match md_path with
          | Some path ->
              Some
                {
                  status = 200;
                  headers =
                    headers_for md_content_type
                      [
                        ( "Link",
                          "<" ^ path
                          ^ ">; rel=\"alternate\"; type=\"text/markdown\", </llms.txt>; rel=\"describedby\""
                        );
                      ];
                  body = store.read path;
                }
          | None ->
              if Accept.preferred_type accept [ Accept.html_type ] = None then
                Some
                  {
                    status = 406;
                    headers =
                      headers_for plain_content_type
                        [ ("Cache-Control", "no-store") ];
                    body = Accept.not_acceptable_body (requested accept);
                  }
              else html_result store html_path md_path
        else html_result store html_path md_path
